contracts/AnybodyProblem.sol

```diff
@@ -0,0 +1,640 @@
+                                              storedOutflightMissile[1] == newInflightMissile[1] &&
+                                              storedOutflightMissile[2] == newInflightMissile[2] &&
+                                              storedOutflightMissile[3] == newInflightMissile[3] &&
+                                              storedOutflightMissile[4] == newInflightMissile[4];
+
+        bool newMissile = newInflightMissile[0] == 0 && newInflightMissile[1] == windowWidth;
+        require(newMissile || matchesStoredOutflightMissile, "Invalid inflightMissile");
+    }
+    // update inflightMissile with new outflight missile
+    uint256[5] memory newOutflightMissile = [
+        input[0],
+        input[1],
+        input[2],
+        input[3],
+        input[4]
+    ];
+    runs[runId].levels[levelIndex].tmpInflightMissile = newOutflightMissile;
+
+    uint256 time = input[5 + bodyCount * 5];
+
+    verifyProof(bodyCount, verifier, a, b, c, input);
+
+    Level memory levelData = runs[runId].levels[levelIndex];
+
+    levelData.time += time;
+    require(levelData.time <= maxTick, "Time limit exceeded");
+
+    uint256 bodiesGone;
+    Body memory bodyData;
+    for (uint256 i = 0; i < bodyCount; i++) {
+        bodyData = levelData.tmpBodyData[i];
+
+        verifyBodyDataMatches(bodyData, input, bodyCount, i);
+        bodyData = extractBodyData(bodyData, input, i);
+
+        if (i == 0) {
+          require(bodyData.radius != 0, "You shot the body you should protect");
+        }
+        if (bodyData.radius == 0) {
+            bodiesGone++;
+        }
+        levelData.tmpBodyData[i] = bodyData;
+    }
+    runs[runId].levels[levelIndex] = levelData;
+
+    if (bodiesGone == level) {
+      runs[runId].levels[levelIndex].solved = true;
+      emit LevelSolved(msg.sender, runId, level, levelData.time, day);
+      runs[runId].accumulativeTime += levelData.time;
+      if (level == LEVELS) {
+        runs[runId].solved = true;
+        require(msg.value == price, "Incorrect payment");
+        (bool sent, bytes memory data) = proceedRecipient.call{
+            value: msg.value
+        }("");
+        Speedruns(speedruns).__mint(runId, msg.sender);
+        emit EthMoved(proceedRecipient, sent, data, msg.value);
+        emit RunSolved(msg.sender, runId, runs[runId].accumulativeTime, day);
+        gamesPlayed[msg.sender].total++;
+        addToLeaderboard(runId);
+
+      } else {
+        addNewLevelData(runId);
+      }
+    }
+  }
+  // TODO: Make sure this is commented out before deployment
+  function testFunctionCommentBeforeDeployment(uint256 runId, uint256 levelIndex, uint256[5] memory missile, Body[6] memory bodyData) public onlyOwner {
+    runs[runId].levels[levelIndex].tmpInflightMissile = missile;
+    runs[runId].levels[levelIndex].tmpBodyData = bodyData;
+  }
+  function addToLeaderboard(uint256 runId) internal {
```

Comment:
this is cool. what do we need the leaderboard api for with this? I guess you'd still need to query events for a user's own stats.

Reply:
in general we want to avoid requiring users to query the blockchain directly. It's slow and expensive with API costs. Better to have them query our server which has already queried the blockchain. Good to have this data on chain anyway too tho because then we can use it programmatically for things like bounties, rewards, etc.

We could emit an event from contract that tells server to update high scores. This would prevent server from having to calculate it and also prevent users from querying blockchain directly for it.

Still haven't implemented weekly speed on-chain. It might be tough because you have to maintain a list of everyone's scores regardless of whether they're a daily winner since it's possible you could have fastest weekly score but not have shown up on one or more daily boards.